subsys/bluetooth/host/hci_core.c

```diff
@@ -193,7 +193,8 @@ void bt_hci_host_num_completed_packets(struct net_buf *buf)
 		return;
 	}
 
-	if (!bt_conn_is_handle_valid(conn)) {
+	if (conn->state != BT_CONN_CONNECTED &&
+	    conn->state != BT_CONN_DISCONNECT) {
```

Comment:
Some controllers may wait with raising the DISCONNECTED_EVT until all received data buffers are forwarded to the host. 
Not sending HCI Host Num Completed command may result in a deadlock. 

Therefore I believe we should not validate against BT_CONN_DISCONNECT here

Reply:
You mean only this ?

``` C
if (conn->state != BT_CONN_CONNECTED
```

Reply:
@Thalley Please take a look.

Reply:
> You mean only this ?
> 
> ```c
> if (conn->state != BT_CONN_CONNECTED
> ```

@LingaoM, yes. It does not make a difference if the host tries to be clever. It anyways need to handle the case where the controller returns an error code

Reply:
> Some controllers may wait with raising the DISCONNECTED_EVT until all received data buffers are forwarded to the host.
Not sending HCI Host Num Completed command may result in a deadlock.

But my understanding is that if we do not send host complete in this state(BT_CONN_DISCONNECT), will it cause a deadlock?

So there is nothing wrong with my modification.

Reply:
I am not 100% sure I follow @rugeGerritsen comments. Ruge, can you elaborate a bit?

Reply:
This is a scenario that can happen:

1. Host calls _HCI_Host_Buffer_Size_ and informs the controller it has one available buffer.
2. Controller receives data from the peer device and sends it to the host. From now on the controller cannot forward data to the host until it has called _HCI_Host_Number_Of_Completed_Packets_.
3. The controller receives another data packet from the peer device. It waits with forwarding it to the host.
4. The host calls _HCI_Disconnect_. The controller waits with forwarding the _HCI_Disconnection_Complete_ event as it still has more data to the host.
5. The host calls _HCI_Host_Number_Of_Completed_Packets_. The controller now forwards the pending data. Then it raises the _HCI_Disconnection_Complete_ event.

Therefore, if we make the host stop sending _HCI_Host_Number_Of_Completed_Packets_ after the  _HCI_Disconnect_ command, the host will never receive the _HCI_Disconnection_Complete_ event.

Reply:
Under my current modification, even in this state(`BT_CONN_DISCONNECT`), which cause by host call `bt_conn_disconnect`, `HCI_Host_Number_Of_Completed_Packets` can be sent.

Reply:
Agree with @LingaoM : In the change here, the host will continue to send completed packets in the BT_CONN_CONNECTED and BT_CONN_DISCONNECT states.

Reply:
Thanks, you are right.